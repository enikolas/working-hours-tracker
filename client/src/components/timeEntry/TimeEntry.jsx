import React from 'react';
import PropTypes from 'prop-types';

import MonthlyCalendar from '../ui/MonthlyCalendar';
import LabourStatistics from '../ui/LabourStatistics';
import FullScreenSpinner from '../ui/FullScreenSpinner';
import AlertModal from '../ui/modals/AlertModal';

import TimeEntryForm from './TimeEntryForm';
import SpecialDayPanel from './SpecialDayPanel';
import CurrentDate from './CurrentDate';

import { Entries } from '../../PropTypes';
import strings from '../../utils/strings';

const TimeEntry = ({
  entries,
  mode,
  selectedDate,
  selectedEntry,
  statistics,
  successMessage,
  errorMessage,
  isPersisted,
  isLoading,
  onDateChange,
  onChangeEntry,
  onChangeMode,
  onCloseAlert,
  onSubmit
}) => (
  <React.Fragment>
    <FullScreenSpinner active={isLoading} />

    <div className="column column-nav">
      <MonthlyCalendar
        selectedDate={selectedDate}
        timeEntries={entries}
        onDateChange={onDateChange}
      />
      <LabourStatistics
        dayBalance={statistics.dayBalance}
        weekBalance={statistics.weekBalance}
        totalBalance={statistics.totalBalance}
        contractedTime={statistics.contractedTime}
        weekDay={statistics.weekDay}
      />
    </div>
    <div className="column column-actions">
      <CurrentDate selectedDate={selectedDate} />
      <main>
        {(selectedEntry.isHoliday || selectedEntry.isVacation) ?
          <SpecialDayPanel
            entry={selectedEntry}
          />
          :
          <TimeEntryForm
            mode={mode}
            entry={selectedEntry}
            isDisabled={false}
            isPersisted={isPersisted}
            successMessage={successMessage}
            errorMessage={errorMessage}
            onChangeEntry={onChangeEntry}
            onChangeMode={onChangeMode}
            onSubmit={onSubmit}
          />
        }
      </main>
    </div>
    <AlertModal
      title={strings.success}
      active={Boolean(successMessage)}
      content={successMessage}
      onClose={onCloseAlert}
      type="success"
    />
    <AlertModal
      title={strings.error}
      active={Boolean(errorMessage)}
      content={errorMessage}
      onClose={onCloseAlert}
      type="error"
    />
  </React.Fragment>
);

TimeEntry.propTypes = {
  entries: PropTypes.arrayOf(Entries),
  mode: PropTypes.string,
  selectedDate: PropTypes.object,
  selectedEntry: Entries,
  statistics: PropTypes.shape({
    dayBalance: PropTypes.number,
    weekBalance: PropTypes.number,
    totalBalance: PropTypes.number,
    contractedTime: PropTypes.number,
    weekDay: PropTypes.number
  }),
  successMessage: PropTypes.string,
  errorMessage: PropTypes.string,
  isPersisted: PropTypes.bool,
  isLoading: PropTypes.bool,
  onDateChange: PropTypes.func,
  onChangeEntry: PropTypes.func,
  onChangeMode: PropTypes.func,
  onCloseAlert: PropTypes.func,
  onSubmit: PropTypes.func
};

TimeEntry.defaultProps = {
  entries: [{}],
  mode: '',
  selectedDate: {},
  selectedEntry: {},
  statistics: {},
  successMessage: '',
  errorMessage: '',
  isPersisted: false,
  isLoading: false,
  onDateChange: () => {},
  onChangeEntry: () => {},
  onChangeMode: () => {},
  onCloseAlert: () => {},
  onSubmit: () => {}
};

export default TimeEntry;
