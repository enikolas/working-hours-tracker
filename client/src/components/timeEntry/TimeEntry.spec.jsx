import React from 'react';
import { shallow } from 'enzyme';
import moment from 'moment-timezone';


import TimeEntry from './TimeEntry';

const defaultProps = {
	entries: [{}],
	selectedDate: moment.parseZone('2018-04-03T11:22:06.038-03:00'),
	selectedEntry: {},
	statistics: {
		dayBalance: 0,
		weekBalance: 0,
		totalBalance: 0,
		contractedTime: 0,
		weekDay: 0
	},
	successMessage: '',
	errorMessage: '',
	isPersisted: false,
	isLoading: false,
	onDateChange: () => {},
	onChangeEntry: () => {},
	onSubmit: () => {}
};

describe('TimeEntry', () => {
	describe('render', () => {
		it('should render MonthlyCalendar and LabourStatistics', () => {
			const wrapper = shallow(<TimeEntry
				{...defaultProps}
			/>);
			expect(wrapper).toMatchSnapshot();
		});
	});
});

