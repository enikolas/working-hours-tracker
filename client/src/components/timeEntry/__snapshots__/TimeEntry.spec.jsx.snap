// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimeEntry render should render MonthlyCalendar and LabourStatistics 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TimeEntry
    entries={
      Array [
        Object {},
      ]
    }
    errorMessage=""
    isLoading={false}
    isPersisted={false}
    mode=""
    onChangeEntry={[Function]}
    onChangeMode={[Function]}
    onCloseAlert={[Function]}
    onDateChange={[Function]}
    onSubmit={[Function]}
    selectedDate={"2018-04-03T14:22:06.038Z"}
    selectedEntry={Object {}}
    statistics={
      Object {
        "contractedTime": 0,
        "dayBalance": 0,
        "totalBalance": 0,
        "weekBalance": 0,
        "weekDay": 0,
      }
    }
    successMessage=""
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <FullScreenSpinner
          active={false}
        />,
        <div
          className="column column-nav"
        >
          <MonthlyCalendar
            onDateChange={[Function]}
            selectedDate={"2018-04-03T14:22:06.038Z"}
            timeEntries={
              Array [
                Object {},
              ]
            }
          />
          <LabourStatistics
            contractedTime={0}
            dayBalance={0}
            totalBalance={0}
            weekBalance={0}
            weekDay={0}
          />
        </div>,
        <div
          className="column column-actions"
        >
          <CurrentDate
            selectedDate={"2018-04-03T14:22:06.038Z"}
          />
          <main>
            <TimeEntryForm
              entry={Object {}}
              errorMessage=""
              isDisabled={false}
              isPersisted={false}
              mode=""
              onChangeEntry={[Function]}
              onChangeMode={[Function]}
              onSubmit={[Function]}
              successMessage=""
            />
          </main>
        </div>,
        <AlertModal
          active={false}
          content=""
          onClose={[Function]}
          title="üëç"
          type="success"
        />,
        <AlertModal
          active={false}
          content=""
          onClose={[Function]}
          title="Erro!"
          type="error"
        />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "active": false,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <MonthlyCalendar
              onDateChange={[Function]}
              selectedDate={"2018-04-03T14:22:06.038Z"}
              timeEntries={
                Array [
                  Object {},
                ]
              }
            />,
            <LabourStatistics
              contractedTime={0}
              dayBalance={0}
              totalBalance={0}
              weekBalance={0}
              weekDay={0}
            />,
          ],
          "className": "column column-nav",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "onDateChange": [Function],
              "selectedDate": "2018-04-03T14:22:06.038Z",
              "timeEntries": Array [
                Object {},
              ],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "contractedTime": 0,
              "dayBalance": 0,
              "totalBalance": 0,
              "weekBalance": 0,
              "weekDay": 0,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <CurrentDate
              selectedDate={"2018-04-03T14:22:06.038Z"}
            />,
            <main>
              <TimeEntryForm
                entry={Object {}}
                errorMessage=""
                isDisabled={false}
                isPersisted={false}
                mode=""
                onChangeEntry={[Function]}
                onChangeMode={[Function]}
                onSubmit={[Function]}
                successMessage=""
              />
            </main>,
          ],
          "className": "column column-actions",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "selectedDate": "2018-04-03T14:22:06.038Z",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <TimeEntryForm
                entry={Object {}}
                errorMessage=""
                isDisabled={false}
                isPersisted={false}
                mode=""
                onChangeEntry={[Function]}
                onChangeMode={[Function]}
                onSubmit={[Function]}
                successMessage=""
              />,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "entry": Object {},
                "errorMessage": "",
                "isDisabled": false,
                "isPersisted": false,
                "mode": "",
                "onChangeEntry": [Function],
                "onChangeMode": [Function],
                "onSubmit": [Function],
                "successMessage": "",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": "main",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "active": false,
          "content": "",
          "onClose": [Function],
          "title": "üëç",
          "type": "success",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "active": false,
          "content": "",
          "onClose": [Function],
          "title": "Erro!",
          "type": "error",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <FullScreenSpinner
            active={false}
          />,
          <div
            className="column column-nav"
          >
            <MonthlyCalendar
              onDateChange={[Function]}
              selectedDate={"2018-04-03T14:22:06.038Z"}
              timeEntries={
                Array [
                  Object {},
                ]
              }
            />
            <LabourStatistics
              contractedTime={0}
              dayBalance={0}
              totalBalance={0}
              weekBalance={0}
              weekDay={0}
            />
          </div>,
          <div
            className="column column-actions"
          >
            <CurrentDate
              selectedDate={"2018-04-03T14:22:06.038Z"}
            />
            <main>
              <TimeEntryForm
                entry={Object {}}
                errorMessage=""
                isDisabled={false}
                isPersisted={false}
                mode=""
                onChangeEntry={[Function]}
                onChangeMode={[Function]}
                onSubmit={[Function]}
                successMessage=""
              />
            </main>
          </div>,
          <AlertModal
            active={false}
            content=""
            onClose={[Function]}
            title="üëç"
            type="success"
          />,
          <AlertModal
            active={false}
            content=""
            onClose={[Function]}
            title="Erro!"
            type="error"
          />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "active": false,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <MonthlyCalendar
                onDateChange={[Function]}
                selectedDate={"2018-04-03T14:22:06.038Z"}
                timeEntries={
                  Array [
                    Object {},
                  ]
                }
              />,
              <LabourStatistics
                contractedTime={0}
                dayBalance={0}
                totalBalance={0}
                weekBalance={0}
                weekDay={0}
              />,
            ],
            "className": "column column-nav",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "onDateChange": [Function],
                "selectedDate": "2018-04-03T14:22:06.038Z",
                "timeEntries": Array [
                  Object {},
                ],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "contractedTime": 0,
                "dayBalance": 0,
                "totalBalance": 0,
                "weekBalance": 0,
                "weekDay": 0,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <CurrentDate
                selectedDate={"2018-04-03T14:22:06.038Z"}
              />,
              <main>
                <TimeEntryForm
                  entry={Object {}}
                  errorMessage=""
                  isDisabled={false}
                  isPersisted={false}
                  mode=""
                  onChangeEntry={[Function]}
                  onChangeMode={[Function]}
                  onSubmit={[Function]}
                  successMessage=""
                />
              </main>,
            ],
            "className": "column column-actions",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "selectedDate": "2018-04-03T14:22:06.038Z",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <TimeEntryForm
                  entry={Object {}}
                  errorMessage=""
                  isDisabled={false}
                  isPersisted={false}
                  mode=""
                  onChangeEntry={[Function]}
                  onChangeMode={[Function]}
                  onSubmit={[Function]}
                  successMessage=""
                />,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "entry": Object {},
                  "errorMessage": "",
                  "isDisabled": false,
                  "isPersisted": false,
                  "mode": "",
                  "onChangeEntry": [Function],
                  "onChangeMode": [Function],
                  "onSubmit": [Function],
                  "successMessage": "",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": "main",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "active": false,
            "content": "",
            "onClose": [Function],
            "title": "üëç",
            "type": "success",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "active": false,
            "content": "",
            "onClose": [Function],
            "title": "Erro!",
            "type": "error",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
